File: drills/ts/solutions/nc60_plus_one.ts
export function plusOne(digits: number[]): number[] {
  // Adding one is a carry that starts at 1 and dies as soon as a digit below
  // nine absorbs it. The only interesting case is when it never does, and the
  // number grows a digit.
  const out: number[] = [];
  let carry = 1;
  for (let i = digits.length - 1; i >= 0; i--) {
    const total = digits[i] + carry;
    out.push(total % 10);
    carry = Math.floor(total / 10);
  }
  if (carry) out.push(carry);
  return out.reverse();
}

File: drills/go/solutions/nc60_plus_one.go
package main

func plusOne(digits []int) []int {
	result := append([]int{}, digits...)
	// Carry from the right; a 9 becomes 0 and the carry moves on. If the
	// carry survives every digit, the number grew a digit.
	for i := len(result) - 1; i >= 0; i-- {
		if result[i] < 9 {
			result[i]++
			return result
		}
		result[i] = 0
	}
	return append([]int{1}, result...)
}
